packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/internal/AppImpl.kt

```diff
@@ -43,21 +42,27 @@ internal class AppImpl(
     )
 
     override suspend fun login(credentials: Credentials): User {
-        return suspendCoroutine { continuation ->
-            RealmInterop.realm_app_log_in_with_credentials(
-                nativePointer,
-                Validation.checkType<CredentialImpl>(credentials, "credentials").nativePointer,
-                object : CinteropCallback {
-                    override fun onSuccess(pointer: NativePointer) {
-                        continuation.resume(UserImpl(pointer, this@AppImpl))
-                    }
+        // suspendCoroutine doesn't allow freezing callback capturing continuation
+        // ... and cannot be resumed on another thread (we probably also want to guarantee that we
+        // are resuming on the same dispatcher), so run our own implementation using a channel
+        val credentials =
+            Validation.checkType<CredentialImpl>(credentials, "credentials").nativePointer
+        val result = Channel<Result<User>>(1)
+        // Should we do this on the users context or an internal dispatcher
+        RealmInterop.realm_app_log_in_with_credentials(
+            nativePointer,
+            credentials,
+            object : CinteropCallback {
```

Comment:
We are going to replicate this pattern all over, with only difference between the mapping of onSuccess, would be nice if we could somehow extract this into a helper class or method

Reply:
I have tried to address this with helper method that builds the callback from a channel. Combined with `Channel<T>.autoClose` it becomes reasonable condense: https://github.com/realm/realm-kotlin/pull/469/files#diff-8f52344e1cfad3228c6bb6989f5d253af6127732d6bf1ed8dcc90103e11f6038R52

We could try to combine the channel pattern with the callback to make it even shorter, but I could only come up with 
```
suspend fun <T> appCall(success: (NativePointer) -> T, appCall: (CinteropCallback) -> Unit): T {
    return Channel<Result<User>>(1).autoClose { channel ->
        appCall(channelResultCallback(channel, success))
        channel.receive().getOrThrow()
    }
}
```
It requires multiple lambda arguments which in my opinion does not make it more readable or easier to write
```
appCall(
    { pointer -> UserImpl(pointer, this@AppImpl) },
    { callback -> 
        RealmInterop.realm_app_log_in_with_credentials(
            nativePointer,
            credentials,
            callback
    )}
)
```
I would say `autoClose` and `channelResultCallback` does a decent job until we have more use cases.